docs/adr/0004-workflow-instance-indexing-and-trigger-acceptance.md

```diff
@@ -0,0 +1,61 @@
+- Release the lease, allowing any trigger(s) raised by the state entry actions to be processed
+
+### Changes resulting from the move to event sourcing
+
+The primary challenge from the move to event sourcing is that we no longer have a "workflow instance store" that we can query to match trigger `Subjects` against workflow instance `Interests`. Instead, we have an event store which is not queryable in the way we need.
+
+The "normal" approach to dealing with this when implementing event sourcing is to have separate processes which read committed events from the store and use these to construct and store projections of the data. In our case, the projection we need is an index of workflow instances and their `Interests`. However, this approach introduces a delay between events being written to the store and our index projections being up to date - generally referred to as "eventual consistency".
+
+Eventual consistency has, ultimately, always been something clients needed to take account of when sending events into the engine. Whilst the message ingestion endpoint uses the long running operations pattern to allow a client to know when processing a trigger has completed, this does not give the client visibility of any side effects of processing that trigger - e.g. further triggers being raised. As a result, subsequent triggers intended for the same instance may not be matched against that instance depending on the outcome of any additional processing.
+
+However, we now have the situation that even when the event has been processed, the projection used for matching triggers to instances may not have been updated.
+
+In the scenario where a trigger targets a number of workflow instances based on their interests, the change in approach does not introduce any new problems; there is currently, and will continue to be, a challenge for the client to ensure that they take into account the type of consistency issues described above.
+
+In the scenario where a trigger is intended to target a specific workflow instance using its Id, it doesn't actually matter whether the index is stale or not as long as it's been created. This is because a workflow instance Id is immutable and will always be part of the interests collection so will always be in the index.
+
+The only scenario that is problematic is (as before) during workflow instance creation. We can break this down into two:
+- a trigger is raised by an action executed as part of workflow initialization
+- a client needs to wait for a workflow instance creation request to complete and then immediately send a trigger.
+
+In the first case, we need to ensure the index entry for the new instance has been created prior to raising the new triggers. The only way of making this the case is to make updating the index and raising the triggers part of the same process. To achieve this, we can implement a way for an action to return a trigger for the workflow engine to create at the appropriate point.
```

Comment:
I think you need **Case 1** and **Case 2** headings to clearly distinguish what you mean by case 1 and case 2.

I thought you were referring to what I now realise are _Case 2 Scenario A_ and _Case 2 Scenario B_ and I got a bit lost in the logic.